pyro/distributions/sine_bivariate_von_mises.py

```diff
@@ -0,0 +1,280 @@
+
+
+def log_I1(orders: int, value: torch.Tensor, terms=250):
+    r""" Compute first n log modified bessel function of first kind
+    .. math ::
+
+        \log(I_v(z)) = v*\log(z/2) + \log(\sum_{k=0}^\inf \exp\left[2*k*\log(z/2) - \sum_kk^k log(kk)
+        - \lgamma(v + k + 1)\right])
+
+    :param orders: orders of the log modified bessel function.
+    :param value: values to compute modified bessel function for
+    :param terms: truncation of summation
+    :return: 0 to orders modified bessel function
+    """
+    orders = orders + 1
+    if len(value.size()) == 0:
+        vshape = torch.Size([1])
+    else:
+        vshape = copy(value.shape)
+    value = value.reshape(-1, 1)
+
+    k = torch.arange(terms, device=value.device)
+    lgammas_all = torch.lgamma(torch.arange(1, terms + orders + 1, device=value.device))
+    assert lgammas_all.shape == (orders + terms,)  # lgamma(0) = inf => start from 1
+
+    lvalues = torch.log(value / 2) * k.view(1, -1)
+    assert lvalues.shape == (_flat_dim(vshape), terms)
+
+    lfactorials = lgammas_all[:terms]
+    assert lfactorials.shape == (terms,)
+
+    lgammas = lgammas_all.repeat(orders).view(orders, -1)
+    assert lgammas.shape == (orders, terms + orders)  # lgamma(0) = inf => start from 1
+
+    indices = k[:orders].view(-1, 1) + k.view(1, -1)
+    assert indices.shape == (orders, terms)
+
+    seqs = (2 * lvalues[None, :, :] - lfactorials[None, None, :] - lgammas.gather(1, indices)[:, None, :]).logsumexp(-1)
+    assert seqs.shape == (orders, _flat_dim(vshape))
+
+    i1s = lvalues[..., :orders].T + seqs
+    assert i1s.shape == (orders, _flat_dim(vshape))
+    return i1s.view(-1, *vshape)
+
+
+class SineBivariateVonMises(TorchDistribution):
+    r""" Unimodal distribution of two dependent angles on the 2-torus (S^1 ⨂ S^1) given by
+
+    .. math::
+
+        C^{-1}\exp(\kappa_1\cos(x-\mu_1) + \kappa_2\cos(x_2 -\mu_2) + \rho\sin(x_1 - \mu_1)\sin(x_2 - \mu_2))
+
+    and
+
+    .. math::
+
+        C = (2\pi)^2 \sum_{i=0} {2i \choose i}
+        \left(\frac{\rho^2}{4\kappa_1\kappa_2}\right)^i I_i(\kappa_1)I_i(\kappa_2),
+
+    where I_i(\cdot) is the modified bessel function of first kind, mu's are the locations of the distribution,
+    kappa's are the concentration and rho gives the correlation between angles x_1 and x_2.
+
+    This distribution is helpful for modeling coupled angles such as torsion angles in peptide chains.
+    To infer parameters, use :class:`~pyro.infer.NUTS` or :class:`~pyro.infer.HMC` with priors that
+    avoid parameterizations where the distribution becomes bimodal; see note below.
+
+    .. note:: Sample efficiency drops as
+
+        .. math::
+
+            \frac{\rho}{\kappa_1\kappa_2} \rightarrow 1
+
+        because the distribution becomes increasingly bimodal.
+
+    .. note:: The correlation and weighted_correlation params are mutually exclusive.
+
+    .. note:: This distribution does not work with :class:`~pyro.infer.svi.SVI`.
```

Comment:
I think you're underselling here.  Maybe say

> In the context of `SVI`, this distribution can be used as a likelihood but not for latent variables.

Even then, I don't understand why it cannot be used for latent variables.  I see that it has a slow sampler and is not reparametrized, but it should still lead to slow-but-correct inference.  Am I missing something?

Reply:
I was under the impression that the gradients estimations would be too noisy to converge to something trustworthy without reparameterization. However, I'm not particularly confident in this assertion.

Reply:
Sounds good, in that case I'd simply say something like
```
.. note:: This distribution does not generate reparameterized samples, and may
    result in slow convergence when used for latent variables in the context of SVI.
```